rust/storage/src/s3.rs

```diff
@@ -169,6 +180,118 @@ impl S3Storage {
+        let part_size = self.download_part_size_bytes as i64;
+        let head_res = self
+            .client
+            .head_object()
+            .bucket(self.bucket.clone())
+            .key(key)
+            .send()
+            .await;
+        let content_length = match head_res {
+            Ok(res) => match res.content_length {
+                Some(len) => len,
+                None => {
+                    panic!("No content length in head response");
+                }
+            },
+            Err(e) => {
+                panic!("Error in head request: {:?}", e);
+            }
+        };
+        // Round up.
+        let num_parts = (content_length as f64 / part_size as f64).ceil() as i64;
+        let mut ranges = Vec::new();
+        for i in 0..num_parts {
+            let start = i * part_size;
+            let end = if i == num_parts - 1 {
+                content_length - 1
+            } else {
+                (i + 1) * part_size - 1
+            };
+            ranges.push((start, end));
+        }
+        (content_length, ranges)
+    }
+
+    pub(super) async fn fetch_range(
+        &self,
+        key: String,
+        range_str: String,
+    ) -> Result<GetObjectOutput, S3GetError> {
+        let res = self
+            .client
+            .get_object()
+            .bucket(self.bucket.clone())
+            .key(key)
+            .range(range_str)
+            .send()
+            .await;
+        match res {
+            Ok(output) => Ok(output),
+            Err(e) => {
+                tracing::error!("Error fetching range: {:?}", e);
+                match e {
+                    SdkError::ServiceError(err) => {
+                        let inner = err.into_err();
+                        match inner {
+                            aws_sdk_s3::operation::get_object::GetObjectError::NoSuchKey(msg) => {
+                                return Err(S3GetError::NoSuchKey(msg.to_string()));
+                            }
+                            aws_sdk_s3::operation::get_object::GetObjectError::InvalidObjectState(msg) => {
+                                return Err(S3GetError::S3GetError(msg.to_string()));
+                            }
+                            aws_sdk_s3::operation::get_object::GetObjectError::Unhandled(_) => {
+                                return Err(S3GetError::S3GetError("unhandled error".to_string()));
+                            }
+                            _ => {
+                                return Err(S3GetError::S3GetError(inner.to_string()));
+                            }
+                        };
+                    }
+                    _ => {
+                        return Err(S3GetError::S3GetError(e.to_string()));
+                    }
+                }
+            }
+        }
+    }
+
+    pub(super) async fn get_parallel(&self, key: &str) -> Result<Arc<Vec<u8>>, S3GetError> {
```

Comment:
I guess we already do this with the regular `.get()`, but for my understanding why do we return an `Arc`?

Reply:
For dedup'ing. If more than one callsite are awaiting the same future then instead of making deep copy of the buffer for both of them we return an `Arc`